### PR TITLE
AirfRANS 4L/256d gc=2.0 — more permissive gradient clipping

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

At the 3L/192d scale, gc=1.5 beat gc=1.0 significantly (0.00935 vs 0.01419). This suggests the 4L/256d architecture may also benefit from less restrictive gradient clipping. We test gc=2.0 (more permissive than gc=1.5 that luffy tests in their joint gc=1.5 + T_max=10 PR) to map the gc sensitivity curve at 4L/256d. With T_max=5 (unchanged), we isolate the contribution of gradient clipping alone. If the model is gradient-starved at gc=1.0, gc=2.0 should show a clear improvement.

## Instructions

Run the 4L/256d config with gc=2.0 (increase from 1.0), T_max=5 unchanged:

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset airfrans --airfrans_task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 2.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb_group airfrans-4L256d-gc
```

Key difference from baseline: `--grad-clip 2.0` (was 1.0); all other params unchanged.

## Baseline

Current best (PR #2727, W&B run: ruurxdqs):
- val_primary/surface_mse: **0.007264** (at epoch 40, hit cap at epoch 50)
- External target: **0.0043**
- Baseline W&B run: [ruurxdqs](https://wandb.ai/wandb/senpai/runs/ruurxdqs)
- Reproduce: `cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py --dataset airfrans --airfrans_task full --optimizer adamw --lr 7e-4 --cosine-t-max 5 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT giyu:

## Results

### Key Metrics

| Metric | Baseline (gc=1.0, PR #2727) | gc=2.0 (this PR) | External Target |
|--------|---------------------------|-------------------|-----------------|
| val_primary/surface_mse (best) | 0.007264 (epoch 40) | **0.002263** (epoch 211) | 0.0043 |
| Δ vs baseline | — | **-68.8%** | — |
| Δ vs external target | — | **-47.4%** | — |
| Epochs below external target | N/A | 37/213 (17.4%) | — |
| Epochs below baseline | N/A | 94/213 (44.1%) | — |
| Total epochs | 50 (hit cap) | 222 (timeout) | — |

### Test Metrics

⚠️ **Test metrics are from the diverged model** (epoch 222, not best checkpoint). The code evaluates the final model state, not the best checkpoint:
- test_primary/surface_mse: 0.5948 (meaningless — model diverged)

**Suggestion**: train.py should reload best checkpoint before final test evaluation.

### Training Dynamics

- **Pre-divergence (epochs 1-213)**: Mean val 0.005207, stdev 0.003081, range 0.016
- **Gradient clipping**: ~175/360 steps clipped per epoch (avg grad_norm_mean 7.89 vs clip=2.0)
- **Catastrophic divergence at epoch 214**: grad_norm_mean exploded from ~4 → 14.9 → 676 → inf over 8 epochs
- Divergence sequence: epoch 213 (0.0048) → 214 (0.072) → 215 (0.210) → 219 (0.533) → 221 (0.599, grad_norm=inf)

### Oscillation Analysis (last 30 pre-diverge epochs, 184-213)

| Stat | Value |
|------|-------|
| Mean | 0.005207 |
| Stdev | 0.003081 |
| Min | 0.002263 |
| Max | 0.018340 |
| Range | 0.016077 |

High oscillation (stdev 59% of mean) — gc=2.0 allows too-large gradient updates, causing wide val swings.

### Command

```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset airfrans --airfrans-task full --optimizer adamw \
  --lr 7e-4 --cosine-t-max 5 --grad-clip 2.0 --weight-decay 1e-2 \
  --no-use-ema --enable-fourier \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --wandb-group airfrans-4L256d-gc \
  --agent giyu \
  --wandb-name "giyu/airfrans-4L256d-gc20"
```

- **W&B run**: [l6eyvld8](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/l6eyvld8)
- **Peak GPU memory**: Not logged (process completed, GPU freed)

### What happened

**gc=2.0 achieves stunning trough values** (0.00226 best, 47% below external target) but with catastrophic instability. The more permissive clipping allows the model to explore deeper minima during cosine annealing troughs, but also permits gradient explosions that eventually caused irreversible divergence at epoch 214.

The hypothesis is confirmed: 4L/256d benefits from relaxing gc beyond 1.0, with gc=2.0 producing much deeper troughs. However, gc=2.0 is too permissive for stable long-duration training — the model eventually catastrophically diverges.

Key insight: **the trough envelope kept deepening right up to divergence** (new best at epoch 211, just 3 epochs before collapse). This suggests the optimal gc value is somewhere between 1.0 and 2.0, close enough to the instability boundary to access deep minima but with enough headroom to avoid divergence.

### Suggested follow-ups

1. **gc=1.5 at 4L/256d** — Sweet spot between gc=1.0 stability and gc=2.0 depth. Luffy's PR tests gc=1.5 + T_max=10, but a gc=1.5 + T_max=5 run would isolate the gc contribution.
2. **Best-checkpoint test evaluation** — train.py should reload the best val checkpoint before final test eval. Without this, divergent runs lose their test metrics.
3. **Early stopping / divergence detection** — If grad_norm_mean exceeds e.g. 10x the running average, stop training and reload best checkpoint for test eval.
4. **gc=1.2 or gc=1.3** — More conservative relaxation that may capture the benefit without the instability risk.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2765 because ADVISOR: Closing as unstable/superseded. W&B confirms val=0.002263 at ep211 before catastrophic divergence at ep214. While the trough reading technically beat the old baseline (0.00277), gc=2.0 is too aggressive — the model diverged and test metric is 0.5948 from the corrupted state, making this an invalid result. Additionally, PR #2771 has now been merged with val=0.001479 as the new baseline, so even the trough reading no longer beats baseline. Strong directional confirmation that relaxed gradient clipping helps AirfRANS, but gc=2.0 overshoots. gc=1.0 and gc=1.5 are being explored in other in-flight experiments as the safer middle ground..
